### PR TITLE
new(sinsp/test): add event based tests

### PR DESCRIPTION
**Pull request by LucaGuerra**
**What type of PR is this?**

/kind design

**Any specific area of the project related to this PR?**

/area tests

**What this PR does / why we need it**:

I would like to propose to add the necessary plumbing to have event-based testing for libsinsp, based on scap events. This PR's goal is to give the reader an idea of what I'd like to do and how this could be accomplished with code in our libs and start a discussion on whether this is useful, how to do this properly, or maybe we want to approach the problem in a different way.

_Update_: people liked the idea and so the PR is now more than a WIP :)

*Goal:*
I propose to have tests to verify the following: given a specific machine initial state and a certain series of events, will the sinsp state be updated correctly? Will Falco or any other client see the expected field for any of the [supported fields for conditions and output](https://falco.org/docs/rules/supported-fields/)? When I patch something inside the parsers, threadinfo and (especially) networking code, can I check if I broke something that was expected (speaking from experience, this happened many times :roll_eyes: ) 

*How:*
Since the latest refactor, libscap gained the feature of having multiple possible syscall sources, neatly divided into engines that can be optionally compiled in the code. I propose to use this feature to:
* Introduce a new module (called `test_input` in the code you see) that will "echo" any event are passed to it and send them straight to sinsp without modification, also optionally able to set a fixed initial state (threads/fds)
* Add a test fixture in libsinsp that will load this engine and can then be used to tell which events to create
* Add tests that will check field values (such as `fd.name`, `proc.name`, ...) potentially after every event to verify that the state has been properly tracked (one example is currently added)

*Why don't you do this with scap files?*

The intelligent reader will have noticed that what I propose is essentially a way to "replay" a bunch of events and an initial state and check a bunch of fields, like Falco rules do. This is done in the official Falco test suite with scap files, so why can't we just do the same? In my opinion:

* Scap files are hard to read for a human and cannot be commented, it's not easy to understand at first glance what they contain and what is the interesting part that should trigger a test
* Scap files are binary, which means they are hard to add / cannot be added to git repos
* Scap files are meant to be read entirely from beginning to end, so it's hard to test the state after a specific event
* Scap files are hard to edit by hand in case we need to test specific corner cases that may be hard to reproduce and record on a live system

**Special notes for your reviewer**:

Do you like the approach? Can you think of a better way? I will add inline comments on some critical parts that we can think about together.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: userspace/libsinsp/test/sinsp.ut.cpp [hunk 1]
```diff
@@ -36,3 +37,91 @@ TEST(sinsp, external_event_processor_initialization)
...
+{
+	add_event(1657881958, 1, PPME_SYSCALL_OPEN_E, 3, "/tmp/the_file", 0, 0);
+
+	open_inspector();
+
+	sinsp_evt *evt;
+	inspector.next(&evt);
+	ASSERT_EQ(evt->get_type(), PPME_SYSCALL_OPEN_E);
+	ASSERT_EQ(get_field_as_string(evt, "evt.arg.name"), "/tmp/the_file");
+}
```
Comment (LucaGuerra) [inline userspace/libsinsp/test/sinsp.ut.cpp:127]: Note for the reviewer: this is the interesting part. The way it works is that:
* every `add_event` call simply has all the fields used for a specific event
* `get_field_as_string` will simply return a field representation of the specific event, so we can check any field we wish (`fd.name`, `proc.name`, ...)
* We can add features to set an initial thread/fd state so we actually have a base to work on

File: userspace/libscap/engine/test_input/test_input.c [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+struct test_input_engine
+{
+	char* m_lasterr;
+	struct scap_test_input_data *m_data;
+	size_t m_event_index;
+};
+
+typedef struct test_input_engine test_input_engine;
+
+#define SCAP_HANDLE_T struct test_input_engine
```
Comment (Andreagit97) [inline userspace/libscap/engine/test_input/test_input.c:33]: Maybe we can split it into a `.h` like for the other engines, WDYT?

File: userspace/libscap/scap_open.h [hunk 3]
```diff
@@ -100,6 +101,8 @@ typedef struct scap_open_args
 
 	scap_source_plugin* input_plugin; ///< use this to configure a source plugin that will produce the events for this capture
 	char* input_plugin_params; ///< optional parameters string for the source plugin pointed by src_plugin
+
+	scap_test_input_data* test_input_data; ///< only used for testing scap consumers by supplying arbitrary test data
```
Comment (Andreagit97) [inline userspace/libscap/scap_open.h:105]: Probably this is not the right PR to change it, but we need to introduce an `enum` to understand which engine we are going to use...

File: userspace/libscap/scap.c [hunk 4]
```diff
@@ -445,6 +445,76 @@ scap_t* scap_open_udig_int(char *error, int32_t *rc,
...
+		scap_close(handle);
+		snprintf(error, SCAP_LASTERR_SIZE, "error copying suppressed comms");
+		return NULL;
+	}
+
+	if(handle->m_vtable->start_capture(handle->m_engine) != SCAP_SUCCESS)
+	{
+		scap_close(handle);
+		return NULL;
+	}
```
Comment (Andreagit97) [inline userspace/libscap/scap.c:514]: Do we really need this? the `noop_start_capture` always return `SCAP_SUCCESS` without doing anything else :thinking:

**Review by Andreagit97 — COMMENTED** [review]
WOW :heart_eyes: I really love this idea, huge work @LucaGuerra thank you very much!! 
Starting from an initial known state and checking that this changes as we want after a bunch of events is the best way we have to assert that `libsinsp` is working correctly, or at least these are my humble thoughts... So big +1 on my side!!

File: userspace/libsinsp/test/sinsp.ut.cpp (hunk 1 above)
Comment (Andreagit97) [reply userspace/libsinsp/test/sinsp.ut.cpp:127]: The possibility to get the specific event field is really powerful and I think this is exactly what we need here :)

File: userspace/libscap/CMakeLists.txt [hunk 5]
```diff
@@ -202,6 +202,10 @@ if(CMAKE_SYSTEM_NAME MATCHES "Linux")
 	add_subdirectory(engine/kmod)
 	target_link_libraries(scap scap_engine_kmod)
 
+	add_definitions(-DHAS_ENGINE_TEST_INPUT)
```
Comment (Molter73) [inline userspace/libscap/CMakeLists.txt:205]: Because this is specifically meant for testing, think we can hide its compilation behind a configuration option like `CREATE_TEST_TARGETS`? Particularly, I think builds meant for release/production should not include this engine.

File: userspace/libsinsp/test/sinsp.ut.cpp (hunk 1 above)
Comment (Andreagit97) [reply userspace/libsinsp/test/sinsp.ut.cpp:127]: Obviously, this is just a test example but I propose to write some comments for every test we add, in which we try to explain the flow we are thinking about and what we expect from this test. I propose this because `libsinsp` internals are quite complex and coming back to test in the future could be not so easy to do without adequate documentation, WDYT about that?

File: userspace/libscap/scap.c (hunk 4 above)
Comment (Molter73) [reply userspace/libscap/scap.c:514]: I don't want to speak for Luca, but personally I have a "better safe than sorry" policy, granted this is just for testing but if the `start_capture` function for this engine is changed to something that can fail, it could lead to some pretty erratic error messages. Just my 2 cents 🤷🏻‍♂️

File: userspace/libsinsp/test/sinsp.ut.cpp [hunk 6]
```diff
@@ -36,3 +37,91 @@ TEST(sinsp, external_event_processor_initialization)
...
+	{
+		inspector.open_test_input(m_test_data.get());
+	}
+
+	scap_evt* add_event(uint64_t ts, uint64_t tid, enum ppm_event_type event_type, uint32_t n, ...)
+	{
+		struct scap_sized_buffer event_buf = {NULL, 0};
+		size_t event_size;
+		char error[SCAP_LASTERR_SIZE];
+		error[0] = '\0';
```
Comment (Molter73) [inline userspace/libsinsp/test/sinsp.ut.cpp:69]: [ultra-nit]
```suggestion
		char error[SCAP_LASTERR_SIZE] = {'\0'};
```

File: userspace/libsinsp/test/sinsp.ut.cpp (hunk 1 above)
Comment (LucaGuerra) [reply userspace/libsinsp/test/sinsp.ut.cpp:127]: Do I read "networking fields"?
Yes absolutely, if the behavior is not obvious (such as opening a file and checking if `fd.name` is right) we can add comments. To make this even more readable we could interleave `add_event()` with `next()` calls as long as the processing is done on a single thread like it is today. That might not be extremely clean but it would show exactly the event we're processing and the expected state change or result.

**Review by Molter73 — COMMENTED** [review]
This is awesome @LucaGuerra! I particularly like that this approach should be extremely repeatable, which should prevent us from having flakes in any tests written. I've left a few comments but overall this is looking great!

File: userspace/libscap/scap.c (hunk 4 above)
Comment (gnosek) [reply userspace/libscap/scap.c:514]: Yes, let's keep it. Somewhere down my patch queue I have one unifying all the various open methods, so the more similar they are, the less ugly it's going to be :)

File: userspace/libscap/scap_open.h (hunk 3 above)
Comment (gnosek) [reply userspace/libscap/scap_open.h:105]: If only we had tagged unions in C... :(

If/when we touch this struct, ideally I'd make it contain a field to identify the engine somehow (yeah, probably an enum), some common stuff (callbacks, suppressed comms etc.) and a single `void*` pointing to whatever the engine wants it to point

**Review by gnosek — COMMENTED** [review]
Nice!

File: userspace/libscap/CMakeLists.txt (hunk 5 above)
Comment (LucaGuerra) [reply userspace/libscap/CMakeLists.txt:205]: I agree, also cmakefiles will need to be modified to remove this if not requested.

File: userspace/libscap/scap_open.h (hunk 3 above)
Comment (Andreagit97) [reply userspace/libscap/scap_open.h:105]: Just one question @gnosek, are you planning to modify it in your vtable PRs or do we want to address it separately?

Comment (gnosek) [reply userspace/libscap/scap_open.h:105]: I think it's better to do it after my open path cleanups so I can do that too.

**Comment by Andreagit97** [comment]
Take a look here https://github.com/falcosecurity/libs/pull/499#issuecomment-1193139432 :)

**Comment by LucaGuerra** [comment]
Last week I received great feedback about this, thank you all, I'll now make this mergeable.
I think I need to add a couple things:
* some refactoring according to reviewers' comments :muscle: 
* A way to configure a minimal initial state (i.e. a `/proc` scan with only one process). I'm honestly a bit nervous about making tests that start from a completely empty state without even an `init(1)` process in the tree because it would mean that we'd exercise what is actually a corner case (events happening without anything in the process tree). The initial process tree state can be picked by each test or custom set or actually set to empty.

When I do I'll remove the wip.
Thanks all again :bow: !

File: userspace/libscap/CMakeLists.txt [hunk 7]
```diff
@@ -142,8 +142,13 @@ endif()
 option(CREATE_TEST_TARGETS "Enable make-targets for unit testing" ON)
 
 if(CREATE_TEST_TARGETS AND NOT WIN32)
-		# Add unit test directories
-		add_subdirectory(test)
+	# Add engine only used for testing
+	add_definitions(-DHAS_ENGINE_TEST_INPUT)
```
Comment (FedeDP) [inline userspace/libscap/CMakeLists.txt:146]: Minor note: i think these tests can be enabled on windows also, right?

**Review by FedeDP — COMMENTED** [review]
Late to the party @LucaGuerra !
This is actually freaking huge! And i really love how it interoperates with existing code leveraging libscap refactoring and `scap_event_encode_params_v` function!

**Comment by LucaGuerra** [comment]
I have updated the PR as mentioned above! Updates are:

* Now we can easily configure the initial state by adding processes and open fds _before_ events are processed
* I also added a few more helper functions to make writing tests easier 😎 
* I changed and commented the first test

File: userspace/libsinsp/test/sinsp.ut.cpp [hunk 8]
```diff
@@ -36,3 +37,22 @@ TEST(sinsp, external_event_processor_initialization)
...
+
+	open_inspector();
+	sinsp_evt *evt;
+
+	// since adding and reading events happens on a single thread they can be interleaved.
+	// tests may need to change if that will not be the case anymore
+	add_event(increasing_ts(), 1, PPME_SYSCALL_OPEN_E, 3, "/tmp/the_file", PPM_O_RDWR, 0);
+	evt = next_event();
+
+	add_event(increasing_ts(), 1, PPME_SYSCALL_OPEN_X, 6, 3, "/tmp/the_file", PPM_O_RDWR, 0, 5, 123);
```
Comment (FedeDP) [inline userspace/libsinsp/test/sinsp.ut.cpp:52]: Looking at the code; is the `next_event` always needed (or, well, most of the time) after the `add_event` call?
If so, can't `add_event` take a `sinsp_evt*` argument and, when not-nil, nex_event gets automatically called?

**Review by FedeDP — COMMENTED** [review]
Left a minor comment, otherwise LGTM!

File: userspace/libsinsp/test/sinsp.ut.cpp (hunk 8 above)
Comment (LucaGuerra) [reply userspace/libsinsp/test/sinsp.ut.cpp:52]: The two approaches are very similar, but I prefer it to be like this to be a bit more flexible:
* currently (like the comment says) you _can_ interleave add_event with next_event, but if that stops being the case you will be required to add all the add_event before the test starts and next_event after.
* you might want to do something fancier when generating events (such as having for loops that generate the same operations for different processes or something like that) and then have a different logic when checking them. Leaving the two operations separate gives you the maximum flexibility I think.
* Another thing we might add if we see that adding `next_event()` is too much of a hassle could be a function that consumes all events that the test generated so far.
* The idea about an optional parameter is cool but IMO it might look a little confusing to a new contributor that wants to add a test

Comment (FedeDP) [reply userspace/libsinsp/test/sinsp.ut.cpp:52]: I have no strong opinion in that sense :) it was just a suggestion i came up with while reading through the code.
:+1: from me!

Comment (FedeDP) [reply userspace/libsinsp/test/sinsp.ut.cpp:52]: We can do something about it later in any case!

**Review by FedeDP — APPROVED** [review]
/approve

File: userspace/libscap/CMakeLists.txt (hunk 7 above)
Comment (Andreagit97) [reply userspace/libscap/CMakeLists.txt:152]: we can do this in another PR maybe :thinking:

Comment (FedeDP) [reply userspace/libscap/CMakeLists.txt:152]: Completely agree!

File: userspace/libsinsp/test/sinsp.ut.cpp (hunk 8 above)
Comment (Andreagit97) [reply userspace/libsinsp/test/sinsp.ut.cpp:52]: Agree with @LucaGuerra here, BTW we can always add something like this if needed

**Review by Andreagit97 — APPROVED** [review]
/approve

**Comment by Andreagit97** [comment]
Amazing job @LucaGuerra !!! :tada:
